site-src/geps/gep-992.md

```diff
@@ -0,0 +1,184 @@
+setting weights on individual objects, and consequently there is no place to
+declare an intended state for other strategies.
+
+It is not clear (via specification) to an end-user what the load-balancing
+implementation for any given `HTTPRoute` will end up being when they post
+an new `HTTPRoute` object.
+
+### Proposed Solution
+
+Route rules can now be provided a `LoadBalancerStrategy`, which is required
+but defaults to the generic `RoundRobinStrategy`:
+
+```golang
+type HTTPRouteRule struct {
+	// <NO documentation updates>
+	Matches []HTTPRouteMatch `json:"matches,omitempty"`
+
+	// <NO documentation updates>
+	Filters []HTTPRouteFilter `json:"filters,omitempty"`
+
+	// BackendRefs defines references to the individual backend(s) where matching
+	// requests should be sent.
+	//
+	// If multiple backends are configured the LoadBalancerStrategy option becomes
+	// required and will define how traffic will be dispersed across those backends.
+	//
+	// If unspecified or invalid (refers to a non-existent resource or a Service
+	// with no endpoints), the rule performs no forwarding. If there are also no
+	// filters specified that would result in a response being sent, a HTTP 503
+	// status code is returned.
+	//
+	// Support: Core for Kubernetes Service
+	// Support: Custom for any other resource
+	//
+	// Support for weight: Core
+	//
+	// +optional
+	// +kubebuilder:validation:MaxItems=16
+	BackendRefs []HTTPBackendRef `json:"backendRefs,omitempty"`
+
+	// LoadBalancerStrategy defines the high level strategy to use when multiple
+	// BackendRefs are configured for traffic.
+	//
+	// If only one backend is provided in BackendRefs this property has no
+	// effect, otherwise it is required.
+	//
+	// Support: Core for RoundRobinStrategy
+	// Support: Core for WeightedStrategy
+	// Support: Extended for LeastConnectionsStrategy
+	//
+	// +optional
+	// +kubebuilder:validation:Enum=RoundRobin,WeightedStrategy,LeastConnectionsStrategy
+	// +kubebuilder:default=RoundRobin
+	LoadBalancerStrategy *LoadBalancerStrategy `json:"loadBalancerStrategy,omitempty"`
+}
+```
+
+Practically speaking this changes very little about how things already work,
+except that end-users and implementers can now very clearly understand the
+intended dispersement of traffic across backends instead of the implicit
+nature of the previous model.
+
+We document the individual strategies and more can be added over time:
+
+```golang
+const (
+	// RoundRobin is a simple LoadBalancerStrategy which distributes inbound
+	// matching requests across all HTTPBackendRefs evenly according to the
+	// overall number of requests.
+	RoundRobinStrategy LoadBalancerStrategy = "RoundRobin"
+
+	// WeightedStrategy distributes inbound matching requests across
+	// HTTPBackendRefs according to the "weight" value set on each individual
+	// ref, and uses those weights to determine what percentage of traffic each
+	// backend should take on.
+	WeightedStrategy LoadBalancerStrategy = "Weighted"
```

Comment:
This is concerning to me because it means that `weight` doesn't apply to other load balancing algorithms.

Its very reasonable to have weight and {RoundRobin,LeastConnections}. As a two tier selection:
* First, pick the 'backend' by weight.
* Next, pick the exact upstream IP to use by LB algorithm

Reply:
I considered this when I originally wrote it, but so far hadn't included it because it seemed simple to add later and I wanted to keep the first draft tiny. If you feel that it really does need to be added up front I can simply add it.

Reply:
This would break us I think, in its current state. We have other resources (eg Policy) that define load balancing state. By not having the strategy optional here, that enforces that we would have to use the Route policy (unless we just blatantly ignore what they configured...), which is very restrictive compared to our current policy options.

Which goes back to my other comment about it being in route, really.